### PR TITLE
Track delayed nodes in qset

### PR DESCRIPTION
**Pull request by marta-lokhova**
this change adds tracking of nodes in the qset that are delayed

### CODE REVIEW

File: src/scp/SCP.h [hunk 1]
```diff
@@ -24,6 +26,8 @@ typedef std::shared_ptr<SCPQuorumSet> SCPQuorumSetPtr;
 class SCP
 {
     SCPDriver& mDriver;
+    std::unordered_map<NodeID, std::shared_ptr<medida::Histogram>>
```
Comment (marta-lokhova) [inline src/scp/SCP.h:29]: I found it convenient to use histograms in order to track nodes voting overtime. However, right now the sliding window defaults to 5 min of data, and I'm wondering if this is too short (we could change medida to allow passing in a longer time window)

**Comment by marta-lokhova** [comment]
example output in `quorum` endpoint:
```
   "lag_ms" : {
      "LOBSTR 1 (Europe)" : 132,
      "LOBSTR 2 (Europe)" : 164,
      "LOBSTR 3 (North America)" : 158,
      "LOBSTR 4 (Asia)" : 389,
      "LOBSTR 5 (Australia)" : 359,
      "SDF 1" : 285,
      "SDF 2" : 271,
      "SDF 3" : 281,
      "SatoshiPay (DE, Frankfurt)" : 163,
      "SatoshiPay (SG, Singapore)" : 307,
      "SatoshiPay (US, Iowa)" : 203,
      "StellarExpert-V1" : 267,
      "StellarExpert-V2" : 238,
      "StellarExpert-V3" : 240
```

File: src/herder/HerderSCPDriver.cpp [hunk 2]
```diff
@@ -976,11 +1018,24 @@ HerderSCPDriver::recordSCPExternalizeEvent(uint64_t slotIndex, NodeID const& id,
...
+        {
+            recordLogTiming(*timing.mSelfExternalize, now,
+                            mSCPMetrics.mExternalizeDelay,
+                            fmt::format("externalize delay ({})",
+                                        mApp.getConfig().toShortString(id)),
+                            std::chrono::nanoseconds::zero(), slotIndex);
+        }
+
+        // Record lag for other nodes
+        auto recordedLag =
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp:1034]: why not just use `medida::Timer` directly and avoid `shared_ptr`, then you can just do
`auto &vLag = mQSetLag[id];`

File: src/herder/HerderSCPDriver.h [hunk 3]
```diff
@@ -202,6 +206,9 @@ class HerderSCPDriver : public SCPDriver
     // Prepare timeouts per ledger
     medida::Histogram& mPrepareTimeout;
 
+    // Externalize lag tracking for nodes in qset
+    std::unordered_map<NodeID, std::shared_ptr<medida::Timer>> mQsetLag;
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.h:210]: should be spelled `mQSetLag`

File: src/herder/HerderImpl.cpp [hunk 4]
```diff
@@ -1187,9 +1187,14 @@ HerderImpl::getJsonQuorumInfo(NodeID const& id, bool summary, bool fullKeys,
     ret["node"] = mApp.getConfig().toStrKey(id, fullKeys);
     ret["qset"] = getSCP().getJsonQuorumInfo(id, summary, fullKeys, index);
     bool isSelf = id == mApp.getConfig().NODE_SEED.getPublicKey();
-    if (isSelf && mLastQuorumMapIntersectionState.hasAnyResults())
+    if (isSelf)
     {
-        ret["transitive"] = getJsonTransitiveQuorumIntersectionInfo(fullKeys);
+        ret["lag_ms"] = getHerderSCPDriver().getQsetLagInfo(summary, fullKeys);
```
Comment (MonsieurNicolas) [inline src/herder/HerderImpl.cpp:1192]: this should be under the `qset` node.

Also, you need to update the `admin.md` file that contains examples (both compact via `info` and normal)

File: src/herder/HerderSCPDriver.cpp [hunk 5]
```diff
@@ -932,6 +932,48 @@ HerderSCPDriver::getPrepareStart(uint64_t slotIndex)
...
+    auto qSet = getSCP().getLocalQuorumSet();
+    LocalNode::forAllNodes(qSet, [&](NodeID const& n) {
+        auto lag = getExternalizeLag(n);
+        if (lag > 0)
+        {
+            if (!summary)
+            {
+                ret[toStrKey(n, fullKeys)] = static_cast<uint64_t>(lag);
+            }
+            totalLag += lag;
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp:951]: from what I see we only need to update those when `summary` is true

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
this looks good to me, only a few minor changes

**Comment by marta-lokhova** [comment]
addressed your feedback @MonsieurNicolas. This has also been squashed/rebased now.

File: docs/software/admin.md [hunk 6]
```diff
@@ -638,6 +639,7 @@ The output will look something like
             "disagree" : 0,
             "fail_at" : 2,
             "hash" : "d5c247",
+            "lag_ms" : 430,
```
Comment (MonsieurNicolas) [inline docs/software/admin.md:642]: you forgot to update the output of the `quorum` in the section `Quorum set diagnostics`

**Review by MonsieurNicolas — CHANGES_REQUESTED** [review]
minor miss in the docs and we can merge it

File: docs/software/admin.md (hunk 6 above)
Comment (marta-lokhova) [reply docs/software/admin.md:642]: ah yes, grepped for the wrong thing. Fixed!

**Comment by MonsieurNicolas** [comment]
r+ a3fca83663f612d5091120ce71ebddd1c50646ac

**Comment by MonsieurNicolas** [comment]
r+ 7f0f5a59cddf11a5e194070a1683f13c6a9d6eae

**Pull request merged**